README.md
# VECTOR DELTA ENGINE DAEMON (VDED)

![VDED](https://github.com/jbuchbinder/vded/raw/master/vded_logo.png)

* Homepage: https://github.com/jbuchbinder/vded
* Twitter: [@jbuchbinder](https://twitter.com/jbuchbinder)

## What it is / What it does

**vded** is a simple REST-type method for tracking deltas of
ever-increasing values (although it handles resets, etc, pretty well).
It is meant to accept submissions of values which increment, then
determine the delta between that value and its predecessor, as well as
figure the rate over time.

It also has the ability to track on/off values, which it refers to as
"switches".

It also de/serializes state to a file, so that history isn't lost when
it restarts.

**vded** is written in [Go](http://golang.org/), for
convenience and (hopefully) speed.

## Building

* Prerequisites (Ubuntu/Debian):
`sudo apt-get install golang`
* Building:
`go build`

## Using

### Control

`http://localhost:48333/control?action=serialize`

Serialize all data to disk in JSON format.

`http://localhost:48333/control?action=shutdown`

Serialize data to disk and shut down VDED service.

### Switches

`http://localhost:48333/switch?host=HOSTNAME&switch=NAME&value=VALUE&ts=TIMESTAMPINSEC&action=put`

* value: ON/on/TRUE/true or OFF/off/FALSE/false
* timestamp: long representation by seconds

`http://localhost:48333/switch?host=HOSTNAME&switch=NAME&action=get`

### Vectors

Queries to **vded** are as simple as
`http://localhost:48333/vector?host=HOSTNAME&vector=NAME&value=VALUE&ts=TIMESTAMPINSEC&submit_metric=TRUEORFALSE&units=UNITS&group=GROUP`

That will submit values, and will return OK if successful.

Dumping the value of a vector can be accomplished with

`http://localhost:48333/dumpvector?host=HOSTNAME&vector=NAME`

which will return a "hash" of values, including:

* `last_diff`: Delta between last data reporting period and this one.
* `per_minute`: Rate per minute since the last piece of data was pushed
   in
* `per_hour`: Rate per hour since the last piece of data was pushed in
* `submit_metric`: (Optional) Whether to enable pushing deltas to
   ganglia through gmetric. Defaults to true. Possible values are:
   TRUE, FALSE, true, false, YES, NO, yes, no, 0, 1
* `units`: (Optional) Unit name used when submitting metrics to Ganglia.
   This defaults to use "count" if nothing is specified.
* `group`: (Optional) Name of Ganglia metrics group. Defaults to using
   "vectors" if nothing is specified.

(Please note that host is optional but the rest of the params aren't, so
you might get an error otherwise.)




debian/changelog
vded (0.0.5) unstable; urgency=low

  * Packaging for Go build.

 -- Jeff Buchbinder <jeff@freemedsoftware.org>  Mon, 25 Jun 2012 14:38:42 -0400

vded (0.0.1) unstable; urgency=low

  * Initial packaging.

 -- Jeff Buchbinder <jeff@freemedsoftware.org>  Mon, 13 Feb 2012 13:21:42 -0500


vded.js
// VDED - Vector Delta Engine Daemon
// https://github.com/jbuchbinder/vded
//
// vim: tabstop=4:softtabstop=4:shiftwidth=4:noexpandtab

var http = require('http');
var url  = require('url');
var fs   = require('fs');
var gm   = require('./gmetric');
var exec = require('child_process').exec;

// Default values, overriden by command arguments
var server_port = 48333;
var pidfile = "";
var statefile = "";
var ganglia_enabled = false;
var ganglia_host = null;
var ganglia_port = 8649;
var ganglia_spoof = null;
var max_entries = 0;
var vectors = {};
var switches = {};
var purgeInterval = 30 * 1000;
var flushInterval = 1 * 60 * 1000;

parseArgs(process.argv.splice(2));

// Write the PID to a file
writePid();

// Load data, if there is any, from state file
deserializeFromFile();

// Initialize gmetric sending, if enabled
var gmetric;
if (ganglia_enabled) {
	gmetric = new gm.gmetric( ganglia_host, ganglia_port, ganglia_spoof != null ? ganglia_spoof : null );
}

http.createServer(function(req,resp) {
	var uparse = url.parse(req.url, true);
	var path = uparse.pathname;
	var args = uparse.query;
	console.log("Request for " + path);
	if (path == '/test') {
		createResponse(resp, 200, ["This is a test",{'a':1,'b':2}]);
	} else if (path == '/favicon.ico') {
		createResponse(resp, 404, "Not found!");
	} else if (path == '/vector' || path == '/submit') {
		var hostname = args['host'] == null ? "localhost" : args['host'];
		var vector_name = args['vector'];
		if (vector_name == null) {
			createResponse(resp, 500, "Bad parameters (vector not given)");
			return;
		}

		var value = args['value'];
		if (value == null) {
			createResponse(resp, 500, "Bad parameters (value not given)");
			return;
		}

		var ts = args['ts'];
		if (ts == null) {
			createResponse(resp, 500, "Bad parameters (ts not given)");
			return;
		}

		var submit_metric = null;
		if (args['submit_metric'] != null) {
			var s = args['submit_metric'];
			if (s.toLowerCase() == 'true' || s == '1' || s.toLowerCase() == 'yes') {
				submit_metric = true;
			} else if (s.toLowerCase() == 'false' || s == '0' || s.toLowerCase() == 'no') {
				submit_metric = false;
			}
		}

		var key = getKeyName(hostname, vector_name);

		if (vectors[key] != null) {
			// Use old entry
			vectors[key].values[ts] = value;
			vectors[key].latest_value = value;
			if (args['spoof'] != null) {
				vectors[key].spoof = args['spoof'];
			}
			if (args['units'] != null) {
				vectors[key].units = args['units'];
			}
			if (args['group'] != null) {
				vectors[key].group = args['group'];
			}
			if (submit_metric != null) {
				vectors[key].submit_metric = submit_metric;
			}
		} else {
			// Create new entry
			var v = {
				'host': hostname,
				'name': vector_name,
				'spoof': args['spoof'] ? args['spoof'] : true,
				'submit_metric': submit_metric == null ? true : submit_metric,
				'latest_value': value,
				'units': args['units'] ? args['units'] : 'count',
				'group': args['group'] ? args['group'] : 'vectors',
				'values': { }
			};
			v.values[ts] = value;
			vectors[key] = v;
		}

		var obj = buildVectorResponse(key);
		//console.log( "Current object value = " + JSON.stringify(obj) );
		//console.log( "obj.values.length = " + Object.keys(obj.values).length );
		if (Object.keys(obj.values).length > 1 && obj.submit_metric) {
			submitToGanglia( obj.host, obj.name, obj, obj['last_diff'] );
			if (obj['per_5min'] != null) {
				submitToGanglia( obj.host, obj.name + '_per_5min', obj, obj['per_5min'] );
			}
			if (obj['per_10min'] != null) {
				submitToGanglia( obj.host, obj.name + '_per_10min', obj, obj['per_10min'] );
			}
			if (obj['per_30min'] != null) {
				submitToGanglia( obj.host, obj.name + '_per_30min', obj, obj['per_30min'] );
			}
			if (obj['per_1hour'] != null) {
				submitToGanglia( obj.host, obj.name + '_per_hour', obj, obj['per_1hour'] );
			}
		}
		createResponse(resp, 200, obj);
	} else if (path == '/switch') {
		var action = args['action'] == null ? "put" : args['action'];
		var hostname = args['host'] == null ? "localhost" : args['host'];
		var switch_name = args['switch'];
		if (switch_name == null) {
			createResponse(resp, 500, "Bad parameters (switch not given)");
			return;
		}
		if (action == 'put') {
			var value = args['value'];
			if (value == null) {
				createResponse(resp, 500, "Bad parameters (value not given)");
				return;
			}
			var actual_value = false;
			if (value.toLowerCase() == 'true' || value.toLowerCase() == 'on') {
				actual_value = true;
			} else if (value.toLowerCase() == 'false' || value.toLowerCase() == 'off') {
				actual_value = false;
			} else {
				createResponse(resp, 500, "Bad value for switch");
				return;
			}

			var ts = args['ts'];
			if (ts == null) {
				createResponse(resp, 500, "Bad parameters (ts not given)");
				return;
			}

			var submit_metric = true;
			if (args['submit_metric'] != null) {
				var s = args['submit_metric'];
				if (s.toLowerCase() == 'true' || s == '1' || s.toLowerCase() == 'yes') {
					submit_metric = true;
				} else if (s.toLowerCase() == 'false' || s == '0' || s.toLowerCase() == 'no') {
					submit_metric = false;
				}
			}

			var key = getKeyName(hostname, switch_name);

			if (switches[key] != null) {
				// Use old entry
				switches[key].values[ts] = actual_value;
				switches[key].latest_value = actual_value;
			} else {
				// Create new entry
				var s = {
					'host': hostname,
					'name': switch_name,
					'submit_metric': submit_metric,
					'latest_value': actual_value,
					'values': { }
				};
				s.values[ts] = actual_value;
				switches[key] = s;
			}
			createResponse(resp, 200, switches[key]);
		} else if (action == 'get') {
			var key = getKeyName(hostname, switch_name);
			if (switches[key] == null) {
				createResponse(resp, 401, "Switch does not exist");
				return;
			}
			createResponse(resp, 200, switches[key]);
			return;
		}
	} else {
		createResponse(resp, 404, "Not found!");
	}
}).listen(server_port);

var flushProcess = setInterval(function () {
	// Save state to file.
	console.log("Serialize to file");
	var timer_begin = Date.now();
	serializeToFile( );
	var timer_end = Date.now();
	console.log("FLUSH|Completed in " + ( timer_end - timer_begin ) + " ms");
}, flushInterval);

// Start up purge process
var purgeProcess = setInterval(function () {
	console.log("PURGE|Start purge process");
	if (max_entries <= 0) {
		console.log("PURGE|No purge limit, exiting.");
		// Skip purging if we have no limit
		return;
	}

	var timer_begin = Date.now();

	for (var i in vectors) {
		var vectorvalues = Object.keys(vectors[i].values).length;
		if (vectorvalues <= max_entries) {
			// If we don't have enough entries, skip this vector
			continue;
		}

		var vKeys = Object.keys(vectors[i].values);
		vKeys.sort();

		// Slice off (NUM_ENTRIES - max_entries) - 1
		vKeys.slice(0, (vectorvalues - max_entries) - 1);
		var idx = 0;
		for (idx = 0; idx < vKeys.length; idx++) {
			var ts = vKeys[idx];
			console.log("PURGE|Vector " + vectors[i].host + "/" + vectors[i].name + " purging ts " + ts);
			delete vectors[i].values[ts];
		}
	}

	var timer_end = Date.now();
	console.log("PURGE|Completed in " + ( timer_end - timer_begin ) + " ms");
}, purgeInterval);

function onExit() {
	// Kill purge process if it's running
	if (purgeProcess != null) {
		console.log("Remove purge process");
		clearInterval( purgeProcess );
	}

	if (flushProcess != null) {
		console.log("Remove flush process");
		clearInterval( flushProcess );
	}

	if (ganglia_enabled) {
		try {
			gmetric.close();
		} catch (e) {
		}
	}

	// Serialize to file on shutdown
	console.log("Serialize to file");
	serializeToFile( );

	// Remove pid
	console.log("Remove PID");
	removePid();

	process.exit();
}

process.on('SIGINT', onExit);
process.on('SIGQUIT', onExit);

// Convenience functions

function parseArgs(argv) {
	var pos=0;
	var curarg='';
	for (pos=0; pos<argv.length; pos++) {
		//console.log(argv[pos]);
		if (curarg != '') {
			// Handle option for argument, if there is one
			switch (curarg) {
				case 'pid':
					pidfile = argv[pos];
					break;
				case 'port':
					server_port = argv[pos];
					break;
				case 'state':
					statefile = argv[pos];
					break;
				case 'ghost':
					ganglia_host = argv[pos];
					ganglia_enabled = true;
					console.log("Enable send to ganglia");
					break;
				case 'gport':
					ganglia_port = argv[pos];
					break;
				case 'gspoof':
					ganglia_spoof = argv[pos];
					break;
				case 'max':
					max_entries = argv[pos];
					break;
				default:
					showSyntax();
					break;
			}
			curarg = '';
		} else {
			switch (argv[pos]) {
				case '-P':
				case '--pid':
					curarg = 'pid';
					break;
				case '-p':
				case '--port':
					curarg = 'port';
					break;
				case '-s':
				case '--state':
					curarg = 'state';
					break;
				case '-G':
				case '--ghost':
					curarg = 'ghost';
					break;
				case '-g':
				case '--gport':
					curarg = 'gport';
					break;
				case '-S':
				case '--gspoof':
					curarg = 'gspoof';
					break;
				case '-m':
				case '--max':
					curarg = 'max';
					break;
				case '-h':
				case '--help':
				default:
					showSyntax();
					break;
			}
		}
	}
	if (curarg != '') {
		showSyntax();
	}
}

function showSyntax() {
	console.log("VDED - Vector Delta Engine Daemon");
	console.log("  -h|--help              Syntax help");
	console.log("  -P|--pid PIDFILE       Path for pid file");
	console.log("  -p|--port PORT         Listening port (default is 48333)");
	console.log("  -s|--state STATEFILE   Path for save state file");
	console.log("  -G|--ghost HOST        Ganglia hostname");
	console.log("  -g|--gport PORT        Ganglia port");
	console.log("  -S|--gspoof IP:HOST    Ganglia default spoof");
	console.log("  -m|--max INT           Maximum number of entries to retain");
	process.exit();
}

function createResponse(resp, status, obj) {
	resp.writeHead(status, {'Content-Type': 'text/plain'});
	resp.write(JSON.stringify(obj));
	resp.end();
}

function buildVectorResponse(key) {
	var timer_begin = Date.now();
	var v = vectors[key];
	if (Object.keys(v.values).length == 1) {
		// Only one, return with no calculation
		v['last_diff'] = v.latest_value;
		v['per_minute'] = 0;
		v['per_hour'] = 0;
	} else {
		// Determine differences
		var keys = new Array();
		for (var i in v['values']) {
			keys.push(i);
		}
		keys.sort();
		var max1 = keys[keys.length - 1];
		var max2 = keys[keys.length - 2];
		var ts_diff = max1 - max2;
		if (ts_diff < 0) { ts_diff = -ts_diff; }
		if (v.values[max1] < v.values[max2]) {
			// Deal with vector value resets, not perfect, but good enough
			v['last_diff'] = v.values[max1];
		} else {
			// As normal, use actual difference
			v['last_diff'] = v.values[max1] - v.values[max2];
		}

		// These to track back in history to find the entries
		// which most closely match their timestamps, and use those
		// values to figure rate over time.
		var key5min  = getClosestAgeEntry(v, v['last_diff'],  300);
		if (key5min != null) {
			var diff = v['last_diff'] - key5min;
			v['per_5min'] = v.values[key5min] / ( diff / 300 );
		}
		var key10min = getClosestAgeEntry(v, v['last_diff'],  600);
		if (key5min != null) {
			var diff = v['last_diff'] - key10min;
			v['per_10min'] = v.values[key10min] / ( diff / 600 );
		}
		var key30min = getClosestAgeEntry(v, v['last_diff'], 1800);
		if (key30min != null) {
			var diff = v['last_diff'] - key10min;
			v['per_30min'] = v.values[key30min] / ( diff / 1800 );
		}
		var key1hour = getClosestAgeEntry(v, v['last_diff'], 3600);
		if (key1hour != null) {
			var diff = v['last_diff'] - key1hour;
			v['per_hour'] = v.values[key1hour] / ( diff / 3600 );
			v['per_1hour'] = v.values[key1hour] / ( diff / 3600 );
		}

		// Standard "per minute" calculation
		v['per_minute'] = (ts_diff < 30) ? 0 : v['last_diff'] / ( ts_diff / 60 );
	}
	var timer_end = Date.now();
	console.log("RESP|Completed in " + ( timer_end - timer_begin ) + " ms");
	return v;
}

function getClosestAgeEntry(vector, now, age) {
	var target = now - age;
	var found = null;
	var bestDiff = 9999999;

	// Grab list of keys (which double as timestamps)
	for (var j in vector.values) {
		if ( Math.abs( j - target ) < bestDiff ) {
			bestDiff = Math.abs( j - target );
			found = j;
		}
	}
	return found;
}

function getKeyName(host, value) {
	return host + '/' + value;
}

function writePid() {
	if (pidfile != '') {
		fs.writeFile( pidfile, process.pid );
	}
}

function removePid() {
	if (pidfile != '') {
		fs.unlinkSync( pidfile );
	}
}

function serializeToFile( ) {
	if (statefile != '') {
		console.log("Serializing state to " + statefile);
		var obj = {};
		obj.vectors = vectors;
		obj.switches = switches;
		var out = JSON.stringify( obj );
		//console.log( out );
		fs.writeFileSync( statefile, out, 'utf8' );
	}
}

function deserializeFromFile() {
	try {
		if (statefile != '' && fs.lstatSync(statefile)) {
			console.log("Retrieving state from " + statefile);
			var raw = fs.readFileSync( statefile, 'utf8' );
			//console.log( raw );
			var obj = JSON.parse( raw );
			vectors = obj['vectors'] != null ? obj['vectors'] : { };
			switches = obj['switches'] != null ? obj['switches'] : { };
		}
	} catch (e) {
		console.log("State file does not exist, skipping deserialization.");
	}
}

function submitToGanglia( host, name, vector, value ) {
	console.log("submitToGanglia()");
	if (!ganglia_enabled) { return; }
	console.log("Send value " + value);

	/*
	// Hack to send using gmetric binary until node-gmetric works properly
	var cmd = "/usr/bin/gmetric " + 
		" -g '" + ( vector.group != null ? vector.group : 'vectors' ) "' " +
		" -n '" + name + "' " +
		" -v '" + value + "' " +
		" -u '" + ( vector.units == null ? 'count' : vector.units ) + "' " +
		" -x 300 -t uint32 " +
		( ganglia_spoof != null ? " -S '" + ganglia_spoof + "'" : "" );
	console.log("GMETRIC CMD: " + cmd);
	exec( cmd, function (error, stdout, stderr) {
		if (error != null) {
			console.log("GMETRIC error  : " + error);
		}
		if (stdout != null || stderr != null) {
			console.log("GMETRIC stdout : " + stdout);
			console.log("GMETRIC stderr : " + stderr);
		}
	});
	*/

	// Native gmetric support
	gmetric.sendMetric( ganglia_spoof != null ? ganglia_spoof : host, name, value, vector.units == null ? 'count' : vector.units, gm.VALUE_INT, gm.SLOPE_BOTH, 300, 300, vector.group == null ? 'vectors' : vector.group );
}

console.log("VDED listening on port " + server_port);



vded.go
// VDED - Vector Delta Engine Daemon
// https://github.com/jbuchbinder/vded
//
// vim: tabstop=4:softtabstop=4:shiftwidth=4:noexpandtab

package main

import (
	"./go-gmetric/gmetric"
	"encoding/json"
	"flag"
	"fmt"
	"log/syslog"
	"net"
	"net/http"
	"os"
	//	"os/signal"
	"sort"
	"strconv"
	"sync"
	//	"syscall"
	"time"
)

var (
	port          = flag.Int("port", 48333, "port to listen for requests")
	state         = flag.String("state", "/var/lib/vded/state.json", "path for save state file")
	ghost         = flag.String("ghost", "localhost", "ganglia host")
	gport         = flag.Int("gport", 8649, "ganglia port")
	spoof         = flag.String("gspoof", "", "ganglia default spoof")
	maxEntries    = flag.Int("max", 300, "maximum number of entries to retain")
	gIP, _        = net.ResolveIPAddr("ip4", *ghost)
	gm            = gmetric.Gmetric{gIP.IP, *gport, *spoof, *spoof}
	log, _        = syslog.New(syslog.LOG_DEBUG, "vded")
	serializeLock *sync.RWMutex
)

type Vector struct {
	Host         string            `json:"host"`
	Name         string            `json:"name"`
	Spoof        bool              `json:"spoof"`
	SubmitMetric bool              `json:"submit_metric"`
	Units        string            `json:"units"`
	Group        string            `json:"group"`
	LatestValue  uint64            `json:"latest_value"`
	LastDiff     uint64            `json:"last_diff"`
	PerMinute    float64           `json:"per_minute"`
	Per5Minutes  float64           `json:"per_5min"`
	Per10Minutes float64           `json:"per_10min"`
	Per30Minutes float64           `json:"per_30min"`
	PerHour      float64           `json:"per_hour"`
	Mutex        *sync.RWMutex     `json:"-"`
	Values       map[string]uint64 `json:"values"`
}

type Switch struct {
	Host        string          `json:"host"`
	Name        string          `json:"name"`
	LatestValue bool            `json:"latest_value"`
	Mutex       *sync.RWMutex   `json:"-"`
	Values      map[string]bool `json:"values"`
}

type SaveState struct {
	Vectors  map[string]*Vector `json:"vectors"`
	Switches map[string]*Switch `json:"switches"`
}

// Store of vectors
var vectors map[string]*Vector
var switches map[string]*Switch

func httpTestHandler(w http.ResponseWriter, r *http.Request) {
	fmt.Fprintf(w, "[\"This is a test\",{\"a\":1,\"b\":2}]")
}

func httpControlHandler(w http.ResponseWriter, r *http.Request) {
	pAction := r.FormValue("action")

	if pAction == "" {
		log.Warning("action was not defined")
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	switch {

	case pAction == "serialize":
		{
			go serializeToFile()
			w.WriteHeader(http.StatusOK)
			fmt.Fprintf(w, "%s", "OK: started serialize job")
		}

	case pAction == "shutdown":
		{
			fmt.Fprintf(w, "%s", "Started shutdown process.")
			log.Warning("[VDED] Shutting down server from control action")
			serializeToFile()
			log.Warning("[VDED] Shutting down NOW")
			os.Exit(0)
		}

	default:
		{
			fmt.Fprintf(w, "%s", "BAD: invalid action")
			w.WriteHeader(http.StatusBadRequest)
		}

	}
}

func httpVectorDumpHandler(w http.ResponseWriter, r *http.Request) {
	//log.Printf("httpVectorDumpHandler()")

	pHost := r.FormValue("host")
	pVector := r.FormValue("vector")

	if pHost == "" || pVector == "" {
		log.Warning("Host and/or vector were not defined")
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	vectorKey := getKeyName(pHost, pVector)

	log.Debug(fmt.Sprintf("Received dump vector request for %s", vectorKey))

	s, err := json.Marshal(vectors[vectorKey])
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}

	w.WriteHeader(http.StatusOK)
	w.Write(s)
}

func httpSwitchHandler(w http.ResponseWriter, r *http.Request) {
	// Grab all proper variables
	pAction := r.FormValue("action")
	pHost := r.FormValue("host")
	pSwitch := r.FormValue("switch")
	pValue := r.FormValue("value")
	pTs := r.FormValue("ts")

	if pAction == "" || pHost == "" || pSwitch == "" {
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	switchKey := getKeyName(pHost, pSwitch)

	switch {
	case pAction == "get":
		{
			if _, ok := switches[switchKey]; ok {
				if switches[switchKey].LatestValue {
					fmt.Fprintf(w, "%s", "true")
				} else {
					fmt.Fprintf(w, "%s", "false")
				}
				return
			} else {
				w.WriteHeader(http.StatusNotFound)
				return
			}
		}
	case pAction == "set":
		{
			value := parseBoolean(pValue, false)
			if _, ok := switches[switchKey]; ok {
				if pTs == "" {
					w.WriteHeader(http.StatusBadRequest)
					return
				}
				if switches[switchKey].Mutex == nil {
					switches[switchKey].Mutex = new(sync.RWMutex)
				}
				switches[switchKey].Mutex.Lock()
				switches[switchKey].Values[pTs] = value
				switches[switchKey].LatestValue = value
				switches[switchKey].Mutex.Unlock()
			} else {
				// Create new switch
				switches[switchKey] = &Switch{
					Host:        pHost,
					Name:        pSwitch,
					LatestValue: value,
					Mutex:       new(sync.RWMutex),
					Values:      make(map[string]bool),
				}
				switches[switchKey].Mutex.Lock()
				switches[switchKey].Values[pTs] = value
				switches[switchKey].Mutex.Unlock()
			}
		}
	default:
		{
			w.WriteHeader(http.StatusBadRequest)
			return
		}
	}

	// TODO: FIXME: XXX: Need to implement switch handling
}

func httpVectorHandler(w http.ResponseWriter, r *http.Request) {
	// Grab all proper variables
	pHost := r.FormValue("host")
	pVector := r.FormValue("vector")
	pValue := r.FormValue("value")
	pTs := r.FormValue("ts")
	pSubmitMetric := parseBoolean(r.FormValue("submit_metric"), true)
	pUnits := r.FormValue("units")
	pSpoof := parseBoolean(r.FormValue("spoof"), true)
	pGroup := r.FormValue("group")

	if pHost == "" || pVector == "" || pValue == "" {
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	// Parse non string values
	value, _ := strconv.ParseUint(pValue, 10, 64)
	vectorKey := getKeyName(pHost, pVector)

	log.Debug(fmt.Sprintf("Received vector request for %s", vectorKey))

	if _, ok := vectors[vectorKey]; ok {
		// Key exists
		if vectors[vectorKey].Mutex == nil {
			vectors[vectorKey].Mutex = new(sync.RWMutex)
		}
		vectors[vectorKey].Mutex.Lock()
		vectors[vectorKey].Values[pTs] = value
		vectors[vectorKey].LatestValue = value
		vectors[vectorKey].Mutex.Unlock()
	} else {
		// Create new vector
		vectors[vectorKey] = &Vector{
			Host:         pHost,
			Name:         pVector,
			SubmitMetric: pSubmitMetric,
			Spoof:        pSpoof,
			Units:        pUnits,
			Group:        pGroup,
			LatestValue:  value,
			LastDiff:     0,
			PerMinute:    0,
			Per5Minutes:  0,
			Per10Minutes: 0,
			Per30Minutes: 0,
			PerHour:      0,
			Mutex:        new(sync.RWMutex),
			Values:       make(map[string]uint64),
		}
		vectors[vectorKey].Mutex.Lock()
		vectors[vectorKey].Values[pTs] = value
		vectors[vectorKey].Mutex.Unlock()
	}

	// Handle building values, async
	go buildVectorKey(vectorKey)

	// Easy response, no data, since we're handling building the values
	// and aggregation asynchronously
	fmt.Fprintf(w, "OK")
}

// Helper functions

func buildVectorKey(key string) {
	// Adjust values
	vectors[key].Mutex.Lock()

	log.Info(fmt.Sprintf("buildVectorKey len = %d", len(vectors[key].Values)))
	if len(vectors[key].Values) <= 1 {
		vectors[key].LastDiff = vectors[key].LatestValue
		vectors[key].PerMinute = 0
		vectors[key].Per5Minutes = 0
		vectors[key].Per10Minutes = 0
		vectors[key].Per30Minutes = 0
		vectors[key].PerHour = 0
	} else {
		keys := make([]string, len(vectors[key].Values))
		i := 0
		for k, _ := range vectors[key].Values {
			keys[i] = k
			i++
		}
		if !sort.StringsAreSorted(keys) {
			sort.Strings(keys)
		}
		max1 := keys[i-1]
		max1int, _ := strconv.ParseUint(max1, 10, 64)
		max2 := keys[i-2]
		max2int, _ := strconv.ParseUint(max2, 10, 64)
		tsDiff := max1int - max2int
		if tsDiff < 0 {
			tsDiff = -tsDiff
		}
		log.Debug(fmt.Sprintf("val1 = %d, val2 = %d", vectors[key].Values[max1], vectors[key].Values[max2]))
		if vectors[key].Values[max1] < vectors[key].Values[max2] {
			// Deal with vector value resets, not perfect, but good enough
			vectors[key].LastDiff = vectors[key].Values[max1]
		} else {
			vectors[key].LastDiff = vectors[key].Values[max1] - vectors[key].Values[max2]
		}

		// TODO: FIXME: Track back in history over time periods

		if tsDiff < 30 {
			vectors[key].PerMinute = 0
		} else {
			vectors[key].PerMinute = float64(vectors[key].LastDiff / (uint64(tsDiff) / 60))
		}
	}

	// TODO: IMPLEMENT: XXX:

	vectors[key].Mutex.Unlock()

	// Submit metric
	go gm.SendMetric(vectors[key].Name, fmt.Sprint(vectors[key].LatestValue), gmetric.VALUE_UNSIGNED_INT, vectors[key].Units, gmetric.SLOPE_BOTH, 300, 600, vectors[key].Group)
}

func getKeyName(hostName, vectorName string) string {
	return hostName + "/" + vectorName
}

func parseBoolean(v string, defaultValue bool) bool {
	if v == "" {
		return defaultValue
	}
	if v == "1" || v == "true" || v == "TRUE" || v == "True" {
		return true
	}
	return false
}

func readState() {
	fi, fierr := os.Stat(*state)
	if fierr != nil {
		log.Err(fierr.Error())
		return
	}

	file, err := os.Open(*state)
	if err != nil {
		log.Err(err.Error())
	}

	data := make([]byte, fi.Size())
	count, err := file.Read(data)
	log.Debug(fmt.Sprintf("Read %d bytes from statefile %s", count, *state))
	if count == 0 {
		// No data read, let's just skip anything else, no fatal errors
		return
	}
	if err != nil {
		file.Close()
		log.Err(err.Error())
	}
	file.Close()

	// Attempt to unmarshal the json data
	var savestate SaveState
	umerr := json.Unmarshal(data, &savestate)
	if umerr != nil {
		log.Err("Could not read data from savestate " + umerr.Error())
	} else {
		vectors = savestate.Vectors
	}

}

func serializeToFile() {
	log.Info("serializeToFile()")

	serializeLock.Lock()
	defer serializeLock.Unlock()

	savestate := &SaveState{
		Vectors:  vectors,
		Switches: switches,
	}

	mTimeStart := time.Now()
	s, err := json.Marshal(savestate)
	if err != nil {
		log.Err(err.Error())
	}
	mTimeEnd := time.Now()

	ioTimeStart := time.Now()
	file, ferr := os.Create(*state)
	if ferr != nil {
		log.Err(ferr.Error())
	} else {
		file.Write(s)
		file.Close()
	}
	ioTimeEnd := time.Now()

	// Get some stats
	mDuration := mTimeEnd.Sub(mTimeStart)
	log.Info(fmt.Sprintf("[SERIALIZE] %s marshalling", mDuration.String()))
	ioDuration := ioTimeEnd.Sub(ioTimeStart)
	log.Info(fmt.Sprintf("[SERIALIZE] %s IO to disk", ioDuration.String()))
}

func handleUdpClient(conn *net.UDPConn) {
	var buf []byte
	buf = make([]byte, 512)
	_, _, err := conn.ReadFromUDP(buf[0:])
	if err != nil {
		return
	}

	log.Info(string(buf))

	data := make(map[string]string)
	jsonerr := json.Unmarshal(buf, &data)
	if jsonerr != nil {
		log.Err(fmt.Sprintf("[UDP] error:", err))
		return
	}

	// TODO: FIXME: XXX: Handle UDP data
}

func udpServer() {
	udpaddr, _ := net.ResolveUDPAddr("udp4", fmt.Sprintf(":%d", *port))
	udpconn, udperr := net.ListenUDP("udp", udpaddr)

	if udperr != nil {
		log.Err(udperr.Error())
		return
	} else {
		for {
			handleUdpClient(udpconn)
		}
	}
}

// Main body

func main() {
	gm.SetLogger(log)
	flag.Parse()

	log.Info("Initializing VDED server")
	vectors = make(map[string]*Vector)
	switches = make(map[string]*Switch)
	serializeLock = new(sync.RWMutex)

	//signalChannel := make(chan os.Signal)
	//signal.Notify(signalChannel, syscall.SIGHUP, syscall.SIGKILL, syscall.SIGQUIT)

	// Read state
	readState()
	defer serializeToFile()

	// Set up gmetric connection

	// Thread to purge old values
	purgeThread := func() {
		log.Info("[PURGE] Thread started")
		for {
			time.Sleep(300 * time.Second)
			for k, _ := range vectors {
				if vectors[k].Mutex == nil {
					vectors[k].Mutex = new(sync.RWMutex)
				}
				vectors[k].Mutex.Lock()
				if len(vectors[k].Values) > *maxEntries {
					targetPurge := len(vectors[k].Values) - *maxEntries
					purgeCount := 0
					for mk, _ := range vectors[k].Values {
						if uint64(purgeCount) < uint64(targetPurge) {
							log.Debug(fmt.Sprintf("[PURGE] %s : %d", k, mk))
							delete(vectors[k].Values, mk)
							purgeCount++
						}
					}
				}
				vectors[k].Mutex.Unlock()
			}
			for sk, _ := range switches {
				if switches[sk].Mutex == nil {
					switches[sk].Mutex = new(sync.RWMutex)
				}
				switches[sk].Mutex.Lock()
				if len(switches[sk].Values) > *maxEntries {
					targetPurge := len(switches[sk].Values) - *maxEntries
					purgeCount := 0
					for mk, _ := range switches[sk].Values {
						if uint64(purgeCount) < uint64(targetPurge) {
							log.Debug(fmt.Sprintf("[PURGE] %s : %d", sk, mk))
							delete(switches[sk].Values, mk)
							purgeCount++
						}
					}
				}
				switches[sk].Mutex.Unlock()
			}
		}
	}
	go purgeThread()

	// Thread to flush state to disk
	flushThread := func() {
		log.Info("[FLUSH] Thread started")
		for {
			time.Sleep(1800 * time.Second)
			serializeToFile()
		}
	}
	go flushThread()

	// Spin up UDP server for requests
	log.Info(fmt.Sprintf("[VDED] Starting UDP service on :%d", *port))
	go udpServer()

	// Spin up HTTP server for requests
	log.Info(fmt.Sprintf("[VDED] Starting HTTP service on :%d", *port))
	httpServer := &http.Server{
		Addr:           fmt.Sprintf(":%d", *port),
		ReadTimeout:    5 * time.Second,
		WriteTimeout:   5 * time.Second,
		MaxHeaderBytes: 1 << 20,
	}
	http.HandleFunc("/test", httpTestHandler)
	http.HandleFunc("/vector", httpVectorHandler)
	http.HandleFunc("/switch", httpSwitchHandler)
	http.HandleFunc("/dumpvector", httpVectorDumpHandler)
	http.HandleFunc("/control", httpControlHandler)
	httpServer.ListenAndServe()

	//sig := <-signalChannel
	//log.Warning("[VDED] Shutting down server with " + sig.String())
	//serializeToFile()
	//os.Exit(0)
}
